core/src/main/scala/org/apache/spark/internal/io/FileCommitProtocol.scala

```diff
@@ -169,4 +184,176 @@ object FileCommitProtocol extends Logging {
+      outputPath: Path): Unit = {
+    val algorithmVersion = getFieldValue(committer, "algorithmVersion").asInstanceOf[Int]
+    val maxAttemptsOnFailure = if (algorithmVersion == 2) {
+      context.getConfiguration.getInt("mapreduce.fileoutputcommitter.failures.attempts", 1)
+    } else {
+      1
+    }
+    var attempt = 0
+    var jobCommitNotFinished = true
+    while (jobCommitNotFinished) {
+      try {
+        commitJobInternal(committer, context, algorithmVersion, outputPath)
+        jobCommitNotFinished = false
+      } catch {
+        case e: Exception =>
+          if (attempt + 1 > maxAttemptsOnFailure) {
+            throw e
+          } else {
+            attempt += 1
+            logWarning(s"Exception get thrown in job commit, retry ($attempt) time.", e)
+          }
+      }
+    }
+    if (committer.getClass.getName.endsWith("ParquetOutputCommitter")) {
+      // Please keep consistent with the implementation of ParquetOutputCommitter.commitJob().
+      var parquetCommitter: Class[_] = null
+      var contextUtil: Class[_] = null
+
+      try {
+        parquetCommitter = Utils.classForName("org.apache.parquet.hadoop.ParquetOutputCommitter")
+        contextUtil = Utils.classForName("org.apache.parquet.hadoop.util.ContextUtil")
+      } catch {
+        case _: ClassNotFoundException =>
+          parquetCommitter = Utils.classForName("parquet.hadoop.ParquetOutputCommitter")
+          contextUtil = Utils.classForName("parquet.hadoop.util.ContextUtil")
+      }
+
+      val getConfiguration = contextUtil.getMethod("getConfiguration", classOf[JobContext])
+      val writeMetadata = parquetCommitter.getMethod("writeMetaDataFile", classOf[Configuration],
+        classOf[Path])
+      val configuration = getConfiguration.invoke(null, context)
+      writeMetadata.invoke(null, configuration, outputPath)
+    }
+  }
+
+  /**
+   * The job has completed, so do following commit job, include:
+   * Move all committed tasks to the final output dir (algorithm 1 only).
+   * Delete the temporary directory, including all of the work directories.
+   * Create a _SUCCESS file to make it as successful.
+   *
+   * Copied from [[FileOutputCommitter]], please keep consistent with it.
+   */
+  @throws[IOException]
+  private def commitJobInternal(
+      committer: FileOutputCommitter,
+      context: JobContext,
+      algorithmVersion: Int,
+      outputPath: Path): Unit = {
+    if (outputPath != null) {
+      val fs = outputPath.getFileSystem(context.getConfiguration)
+
+      if (algorithmVersion == 1) {
+        for (stat <- getAllCommittedTaskPaths(committer, context)) {
+          mergePaths(committer, fs, stat, outputPath)
+        }
+      } else {
+        invokeMethod(committer, "cleanupJob", Seq(classOf[JobContext]), Seq(context))
+        val stagingOutput = new Path(context.getConfiguration.get(FileOutputFormat.OUTDIR))
+        mergePaths(committer, fs, fs.getFileStatus(stagingOutput), outputPath)
```

Comment:
Here, For algorithmVersion=2,  it has committed task output to stagingOutputPath, so I merge paths from stagingOutputPath to tablePath.

I think we don't need implement our own `commitTask` for InsertIntoHadoopFsRelation operation when algorithmVersion=2, which may produce partial result.
@cloud-fan